File: example/java/spring-boot/one_consumer_one_provider/provider/src/test/java/br/com/zup/pact/provider/resource/AccountResourceEndpointTest.java
package br.com.zup.pact.provider.resource;

import static org.hamcrest.Matchers.hasSize;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import br.com.zup.pact.provider.dto.AccountDetailsDTO;
import br.com.zup.pact.provider.dto.BalanceDTO;
import br.com.zup.pact.provider.entity.Account;
import br.com.zup.pact.provider.service.AccountService;
import br.com.zup.pact.provider.stub.AccountStub;
import java.util.Optional;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.web.servlet.MockMvc;

@SpringBootTest
@AutoConfigureMockMvc
class AccountResourceEndpointTest {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private AccountService accountService;

    @Autowired
    private AccountStub accountStub;

    @Test
    void getAccountDetailsByClientId() throws Exception {
        final AccountDetailsDTO accountDetailsDTO = Account.fromEntityToDto(accountStub.getAccounts().get(1));
        when(accountService.getAccountDetailsByClientId(anyInt())).thenReturn(Optional.of(accountDetailsDTO));
        mockMvc.perform(get("/v1/accounts/1"))
                .andDo(print())
                .andExpect(jsonPath("$.accountId").exists())
                .andExpect(jsonPath("$.balance").exists())
                .andExpect(jsonPath("$.accountType").exists())
                .andExpect(jsonPath("$.accountId").value(accountDetailsDTO.getAccountId()))
                .andExpect(jsonPath("$.balance").value(accountDetailsDTO.getBalance()))
                .andExpect(jsonPath("$.accountType").value(accountDetailsDTO.getAccountType().toString()))
                .andExpect(status().isOk());
    }

    @Test
    void getAccountDetailsByNonExistentClientId() throws Exception {
        mockMvc.perform(get("/v1/accounts/1100"))
                .andDo(print())
                .andExpect(status().isNotFound());
    }

    @Test
    void getAll() throws Exception {
        when(accountService.getAll()).thenReturn(Optional.of(accountStub.getAllStubsDTOFormat()));
        final AccountDetailsDTO firstAccountDetailsDTO = accountStub.getAllStubsDTOFormat().get(0);
        mockMvc.perform(get("/v1/accounts"))
                .andDo(print())
                .andExpect(jsonPath("$", hasSize(accountStub.getAllStubsDTOFormat().size())))
                .andExpect(jsonPath("$.[0].accountId").exists())
                .andExpect(jsonPath("$.[0].balance").exists())
                .andExpect(jsonPath("$.[0].accountType").exists())
                .andExpect(jsonPath("$.[0].accountId").value(firstAccountDetailsDTO.getAccountId()))
                .andExpect(jsonPath("$.[0].accountType").value(firstAccountDetailsDTO.getAccountType().toString()))
                .andExpect(status().isOk());
    }

    @Test
    void getAllNoContent() throws Exception {
        when(accountService.getAll()).thenReturn(Optional.empty());
        mockMvc.perform(get("/v1/accounts"))
                .andDo(print())
                .andExpect(status().isNotFound());
    }

    @Test
    void getBalanceOfClientId() throws Exception {
        final BalanceDTO balanceDTO = BalanceDTO.fromAccountToDTO(accountStub.getAccounts().get(1));
        when(accountService.getBalanceByClientId(anyInt())).thenReturn(Optional.of(balanceDTO));
        mockMvc.perform(get("/v1/accounts/balance/1"))
                .andDo(print())
                .andExpect(jsonPath("$.clientId").exists())
                .andExpect(jsonPath("$.accountId").exists())
                .andExpect(jsonPath("$.clientId").value(balanceDTO.getClientId()))
                .andExpect(jsonPath("$.accountId").value(balanceDTO.getAccountId()))
                .andExpect(status().isOk());
    }

    @Test
    void getBalanceOfClientWithNoAccount() throws Exception {
        when(accountService.getBalanceByClientId(anyInt())).thenReturn(Optional.empty());
        mockMvc.perform(get("/v1/accounts/balance/1000"))
                .andDo(print())
                .andExpect(status().isNotFound());
    }
}

File: example/kotlin/one-consumer-one-provider/provider/src/test/kotlin/br/com/zup/pact/provider/resource/AccountResourceEndpointTest.kt
package br.com.zup.pact.provider.resource

import br.com.zup.pact.provider.dto.AccountDetailsDTO
import br.com.zup.pact.provider.dto.BalanceDTO
import br.com.zup.pact.provider.enum.AccountType
import br.com.zup.pact.provider.service.AccountService
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.TestConfiguration
import org.springframework.context.annotation.Bean
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultHandlers
import org.springframework.test.web.servlet.result.MockMvcResultMatchers

@SpringBootTest
@AutoConfigureMockMvc
class AccountResourceEndpointTest(
        @Autowired val mockMvc: MockMvc,
        @Autowired val accountService: AccountService
) {

    @TestConfiguration
    class AccountResourceEndpointTestConfig {
        @Bean
        fun accountService() = mockk<AccountService>()
    }

    @Test
    fun getAccountDetailsByClientId() {
        val accountId = 1
        val balance = 2.0
        val accountType = AccountType.COMMON
        val accountDetailsDTO = AccountDetailsDTO(accountId, balance, accountType)
        every { accountService.getAccountDetailsByClientId(any()) }
                .returns(accountDetailsDTO)

        mockMvc.perform(MockMvcRequestBuilders.get("/v1/accounts/1"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.balance").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountType").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").value(accountId))
                .andExpect(MockMvcResultMatchers.jsonPath("$.balance").value(balance))
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountType").value(accountType.toString()))
                .andExpect(MockMvcResultMatchers.status().isOk)
    }

    @Test
    fun getAccountDetailsByNonExistentClientId() {
        every { accountService.getAccountDetailsByClientId(any()) }
                .returns(null)
        mockMvc.perform(MockMvcRequestBuilders.get("/v1/accounts/1"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.balance").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountType").doesNotExist())
                .andExpect(MockMvcResultMatchers.status().`is`(404))
    }

    @Test
    fun getAll() {
        val accountDetailsDTO1 = AccountDetailsDTO(1, 5.0, AccountType.COMMON)
        val accountDetailsDTO2 = AccountDetailsDTO(2, 6.0, AccountType.COMMON)
        val accountDetailsDTO3 = AccountDetailsDTO(3, 8.0, AccountType.COMMON)
        val accountDetailsDTO4 = AccountDetailsDTO(4, 90.0, AccountType.COMMON)

        val accountDetailsDTOList: List<AccountDetailsDTO> = arrayListOf(
                accountDetailsDTO1, accountDetailsDTO2, accountDetailsDTO3, accountDetailsDTO4)

        every { accountService.getAll() }
                .returns(accountDetailsDTOList)
        mockMvc.perform(MockMvcRequestBuilders.get("/v1/accounts"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.length()").value(accountDetailsDTOList.size))
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].accountId").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].balance").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].accountType").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].accountId").value(accountDetailsDTO1.accountId))
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].balance").value(accountDetailsDTO1.balance))
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].accountType").value(accountDetailsDTO1.accountType.toString()))
                .andExpect(MockMvcResultMatchers.status().isOk)
    }

    @Test
    fun getAllByNonExistentClients() {
        val allAccounts: List<AccountDetailsDTO> = arrayListOf()
        every { accountService.getAll() }.returns(allAccounts)
        mockMvc.perform(MockMvcRequestBuilders.get("/v1/accounts"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].accountId").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].balance").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.[0].accountType").doesNotExist())
                .andExpect(MockMvcResultMatchers.status().`is`(404))
    }

    @Test
    fun getBalanceByClientId() {
        val accountId = 1
        val clientId = 1
        val balance = 10.0
        val balanceDTO = BalanceDTO(accountId, clientId, balance)

        every { accountService.getBalanceByClientId(any()) }
                .returns(balanceDTO)
        mockMvc.perform(MockMvcRequestBuilders.get("/v1/accounts/1/balance"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.clientId").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.balance").exists())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").value(accountId))
                .andExpect(MockMvcResultMatchers.jsonPath("$.clientId").value(clientId))
                .andExpect(MockMvcResultMatchers.jsonPath("$.balance").value(balance))
                .andExpect(MockMvcResultMatchers.status().isOk)
    }

    @Test
    fun getBalanceByNonExistentClientId() {
        every { accountService.getBalanceByClientId(any()) }
                .returns(null)

        mockMvc.perform(MockMvcRequestBuilders.get("/v1/accounts/balance/1"))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.jsonPath("$.accountId").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.clientId").doesNotExist())
                .andExpect(MockMvcResultMatchers.jsonPath("$.balance").doesNotExist())
                .andExpect(MockMvcResultMatchers.status().`is`(404))
    }

}
